pkg/csi/cinder/controllerserver.go

```diff
@@ -147,6 +164,38 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 	return getCreateVolumeResponse(vol, ignoreVolumeAZ, req.GetAccessibilityRequirements()), nil
 }
 
+func (cs *controllerServer) getSnapshot(cloud openstack.IOpenStack, snapshotID string) (error, error) {
+	snap, err := cloud.GetSnapshotByID(snapshotID)
+	// Case: Snapshot exists but is not yet available
+	if err == nil && snap.Status != "available" {
+		return err, status.Errorf(codes.Internal, "Snapshot is not yet available. Status: %s", snap.Status)
+	}
+	// Case: General error while getting the Snapshot
+	if err != nil && !cpoerrors.IsNotFound(err) {
+		return err, status.Errorf(codes.Internal, "Failed to retrieve the snapshot %s: %v", snapshotID, err)
+	}
+
+	return err, nil
+}
+
+func (cs *controllerServer) getBackup(cloud openstack.IOpenStack, snapshotID string) (error, error) {
+	// Backup exists but is not yet available.
+	back, err := cloud.GetBackupByID(snapshotID)
```

Comment:
you don't need to pass the `cloud openstack.IOpenStack` since it's a part of the `cs *controllerServer` struct.
```suggestion
	back, err := cs.Cloud.GetBackupByID(snapshotID)
```

Reply:
while I read this code more, I am increasingly convinced that it is quite possible to get rid of these two methods: `getSnapshot` and `getBackup`. Probably you can combine the logic from them and `CreateVolume`'s one into one method.

Reply:
I deleted both functions, I hope the code is easier to read now